packages/eslint-plugin-react-hooks/src/ReactiveDeps.js

```diff
@@ -0,0 +1,534 @@
+
+      function satisfies(actualDep, dep) {
+        return actualDep === dep || actualDep.startsWith(dep + '.');
+      }
+
+      // TODO: this could use some refactoring and optimizations.
+      // First, ensure what user specified makes sense.
+      declaredDependencies.forEach(({key}) => {
+        if (actualDependencies.some(actualDep => satisfies(actualDep, key))) {
+          // Legit dependency.
+          if (suggestedDependencies.indexOf(key) === -1) {
+            suggestedDependencies.push(key);
+          } else {
+            // Duplicate. Do nothing.
+            duplicateDependencies.add(key);
+          }
+        } else {
+          // Unnecessary dependency. Do nothing.
+          unnecessaryDependencies.add(key);
+        }
+      });
+      // Then fill in the missing ones.
+      dependencies.forEach((info, key) => {
+        if (
+          !suggestedDependencies.some(suggestedDep =>
+            satisfies(key, suggestedDep),
+          )
+        ) {
+          // Legit missing.
+          suggestedDependencies.push(key);
+          if (!info.isKnownToBeStatic) {
+            missingDependencies.add(key);
+          }
+        } else {
+          // Already did that. Do nothing.
+        }
+      });
+
+      const problemCount =
+        duplicateDependencies.size +
+        missingDependencies.size +
+        unnecessaryDependencies.size;
+
+      if (problemCount === 0) {
+        return;
+      }
+
+      const quote = name => "'" + name + "'";
+      const join = (arr, forceComma) => {
+        let s = '';
+        for (let i = 0; i < arr.length; i++) {
+          s += arr[i];
+          if (i === 0 && arr.length === 2) {
+            s += ' and ';
+          } else if (i === arr.length - 2 && arr.length > 2) {
+            s += ', and ';
+          } else if (i < arr.length - 1) {
+            s += ', ';
+          }
+        }
+        return s;
+      };
+      const list = (set, singlePrefix, label, fixVerb) => {
+        if (set.size === 0) {
+          return null;
+        }
+        return (
+          (set.size > 1 ? '' : singlePrefix + ' ') +
+          label +
+          ' ' +
+          (set.size > 1 ? 'dependencies' : 'dependency') +
+          ': ' +
+          join(Array.from(set).map(quote)) +
+          `. Either ${fixVerb} ${
+            set.size > 1 ? 'them' : 'it'
+          } or remove the dependency array.`
+        );
+      };
+      let extraWarning = '';
+      if (unnecessaryDependencies.size > 0) {
+        let badRef = null;
+        Array.from(unnecessaryDependencies.keys()).forEach(key => {
+          if (badRef !== null) {
+            return;
+          }
+          if (key.endsWith('.current')) {
+            badRef = key;
+          }
+        });
+        if (badRef !== null) {
+          extraWarning =
+            ` Mutable values like '${badRef}' aren't valid dependencies ` +
+            "because their mutation doesn't re-render the component.";
```

Comment:
What's an example of code that would fire this warning? The rule supports refs, and e.g. props objects. It seems to be okay with other mutable values you pass it.

Reply:
Not sure I understand what you're saying.

You can see the code causing this in tests, e.g.:

https://github.com/facebook/react/blob/0c722bb0d9d2e2256b0cd29398b3da96f5d20752/packages/eslint-plugin-react-hooks/__tests__/ESLintRuleReactiveDeps-test.js#L1306-L1315

Passing `[ref.current]` in deps is almost always a mistake (which leads to rather gnarly bugs) because it's mutated outside the render phase. So during render it will always be "one step behind".